kanalony-web-server/app/controllers/Application.scala

```diff
@@ -0,0 +1,126 @@
+package controllers
+
+import kanalony.storage.logic.queries.model._
+import kanalony.storage.logic._
+import model._
+import org.joda.time.DateTime
+import scala.concurrent.ExecutionContext.Implicits.global
+import model.Implicits._
+import play.api.mvc._
+import argonaut._
+import Argonaut._
+import scala.concurrent._
+
+class Application extends Controller {
+
+  def query: Action[AnyContent] = Action.async({ implicit request => doQuery(request) })
+
+  def doQuery(request : Request[AnyContent]) : Future[Result] = {
+    try {
+      val analyticsRequestOption = for {
+        input <- request.body.asJson
+        req <- input.toString.decodeOption[AnalyticsRequest]
+      } yield req
+
+      if (analyticsRequestOption.isEmpty) {
+        return wrapWithPromise(BadRequest("Invalid input - ensure proper JSON is supplied with the mandatory fields supplied and that the supplied MIME type is application/json"))
+      }
+
+      val queryParams = requestToQueryParams(analyticsRequestOption.get)
+      val queryExecutionResult = execute(queryParams)
+      val queryResponseData =  queryExecutionResult map { data => AnalyticsResponse(data.headers, data.rows) }
+
+      queryResponseData map {
+        data => Ok(data.asJson.toString)
+      } recover {
+        case e : Exception => BadRequest(e.getMessage)
+      }
+    }
+    catch {
+      case e: QueryNotSupportedException => {
+        return wrapWithPromise(BadRequest(e.getMessage))
+      }
+      case e: IllegalArgumentException => {
+        return wrapWithPromise(BadRequest(e.getMessage))
+      }
+      case e: Exception => {
+        println(e.getClass.getTypeName)
+        return wrapWithPromise(InternalServerError(e.getMessage))
+      }
+    }
+  }
+
+  def wrapWithPromise(res : Result): Future[Result] = {
+    return (Promise[Result] success res).future
+  }
+
+  def execute(queryParams: QueryParams): Future[IQueryResult] = {
+    QueryLocator.locate(queryParams).query(queryParams)
+  }
+
+  def extractValues[T, E <: Exception](values : List[String], enumConverter : String => T, exceptionCreator : String => E): List[T] = {
+    var processedValue = ""
+    var res = List[T]()
+    try {
+      values.foreach(d => {
+        processedValue = d
+        res = res :+ enumConverter(d)
+      })
+      res
+    }
+    catch {
+      case e : Exception => throw exceptionCreator(processedValue)
+    }
+  }
+
+  def extractMetrics(metrics : List[String]): List[Metrics.Value] = {
+    extractValues(metrics, Metrics.withName(_), s => new InvalidMetricsException("Metric " + s + " not supported"))
```

Comment:
Please use string interpolation instead of concatenation:
`metric => new InvalidMetricsException(s"Metric $metric not supported")`